Python 3.14 Android build failures: `_remote_debugging_module` and SDL2 bootstrap (`pyconfig.h` missing)

When building a Python 3.14 project for Android using Buildozer and `python-for-android`(`develop`), two separate compilation failures occur:

1. **`_remote_debugging_module` fails**
   The module fails to compile because `search_linux_map_for_section` is undeclared:

   ```
   /Modules/_remote_debugging_module.c:31: error: call to undeclared function 'search_linux_map_for_section'
   /Modules/_remote_debugging_module.c:814: error: call to undeclared function 'search_linux_map_for_section'
   ```

   This prevents the Python core from building properly for Android.

2. **SDL2 bootstrap fails (`pyconfig.h` not found)**
   After `_remote_debugging_module` compilation fails, attempting to build (`buildozer -v --stacktrace --scan  android debug deploy run`) results in:

   ```
   In file included from start.c:3:
   Python.h:14:10: fatal error: 'pyconfig.h' file not found
      14 | #include "pyconfig.h"
         |          ^~~~~~~~~~~~
   ```

   This appears to be caused by the SDL2 bootstrap not locating Python 3.14 headers correctly.

**Steps to reproduce:**

1. Create a Buildozer project with Python 3.14 requirement.
2. Run:

```bash
buildozer android debug
```

4. Observe compilation errors for `_remote_debugging_module` and `start.c`.

**Environment:**

* OS: Linux
* Buildozer version: latest (from pip)
* python-for-android: latest (from GitHub)
* Python version: 3.14
* NDK: r29
* SDK: Android API 36
* Target architectures: `arm64-v8a`, `armeabi-v7a`

**Additional context:**

* No modifications or patches were applied to Python, Buildozer, or the bootstrap.
* Both issues seem to stem from incomplete support for Python 3.14 in SDL2 bootstrap and `_remote_debugging_module`.

**Request:**

* Guidance or updates to `python-for-android` and SDL2 bootstrap to support Python 3.14 on Android.
* Fix or workaround for `_remote_debugging_module` compilation errors.
* Ensure SDL2 bootstrap can locate Python headers like `pyconfig.h` for Python 3.14 builds.

--- a/pythonforandroid/build.py
+++ b/pythonforandroid/build.py
@@ -694,31 +694,47 @@ def process_python_modules(ctx, modules):
     processed_modules.extend(modules)
 
     if len(modules) == 0:
-        return modules
+        return processed_modules
 
     # temp file for pip report
     fd, path = tempfile.mkstemp()
     os.close(fd)
 
     # setup hostpython recipe
-    host_recipe = Recipe.get_recipe("hostpython3", ctx)
-
     env = environ.copy()
-    _python_path = host_recipe.get_path_to_python()
-    libdir = glob.glob(join(_python_path, "build", "lib*"))
-    env['PYTHONPATH'] = host_recipe.site_dir + ":" + join(
-        _python_path, "Modules") + ":" + (libdir[0] if libdir else "")
+    try:
+        host_recipe = Recipe.get_recipe("hostpython3", ctx)
+        _python_path = host_recipe.get_path_to_python()
+        libdir = glob.glob(join(_python_path, "build", "lib*"))
+        env['PYTHONPATH'] = host_recipe.site_dir + ":" + join(
+            _python_path, "Modules") + ":" + (libdir[0] if libdir else "")
+        pip = host_recipe.pip
+    except Exception:
+        # hostpython3 non available so we use system pip (like in tests)
+        pip = sh.Command("pip")
 
-    shprint(
-        host_recipe.pip, 'install', *modules,
-        '--dry-run', '--break-system-packages', '--ignore-installed',
-        '--report', path, '-q', _env=env
-    )
+    try:
+        shprint(
+            pip, 'install', *modules,
+            '--dry-run', '--break-system-packages', '--ignore-installed',
+            '--report', path, '-q', _env=env
+        )
+    except Exception as e:
+        warning(f"Auto module resolution failed: {e}")
+        return processed_modules
 
     with open(path, "r") as f:
-        report = json.load(f)
+        try:
+            report = json.load(f)
+        except Exception:
+            report = {}
 
     os.remove(path)
+
+    if "install" not in report.keys():
+        # pip changed json reporting format?
+        warning("Auto module resolution failed: invalid json!")
+        return processed_modules
 
     info('Extra resolved pure python dependencies :')
 
@@ -740,7 +756,7 @@ def process_python_modules(ctx, modules):
             pure_python = False
 
         # does this module matches any recipe name?
-        if mname.lower() in _requirement_names:
+        if mname.lower().replace("-", "_") in _requirement_names:
             continue
 
         color = Out_Fore.GREEN if pure_python else Out_Fore.RED
@@ -777,9 +793,7 @@ def run_pymodules_install(ctx, arch, modules, project_dir=None,
 
     info('*** PYTHON PACKAGE / PROJECT INSTALL STAGE FOR ARCH: {} ***'.format(arch))
 
-    # don't run process_python_modules in tests
-    if ctx.recipe_build_order.__class__.__name__ != "Mock":
-        modules = process_python_modules(ctx, modules)
+    modules = process_python_modules(ctx, modules)
 
     modules = [m for m in modules if ctx.not_has_package(m, arch)]
 

--- a/pythonforandroid/build.py
+++ b/pythonforandroid/build.py
@@ -2,9 +2,11 @@ from contextlib import suppress
 import copy
 import glob
 import os
+import json
+import tempfile
 from os import environ
 from os.path import (
-    abspath, join, realpath, dirname, expanduser, exists
+    abspath, join, realpath, dirname, expanduser, exists, basename
 )
 import re
 import shutil
@@ -12,9 +14,12 @@ import subprocess
 
 import sh
 
+from packaging.utils import parse_wheel_filename
+from packaging.requirements import Requirement
+
 from pythonforandroid.androidndk import AndroidNDK
 from pythonforandroid.archs import ArchARM, ArchARMv7_a, ArchAarch_64, Archx86, Archx86_64
-from pythonforandroid.logger import (info, warning, info_notify, info_main, shprint)
+from pythonforandroid.logger import (info, warning, info_notify, info_main, shprint, Out_Style, Out_Fore)
 from pythonforandroid.pythonpackage import get_package_name
 from pythonforandroid.recipe import CythonRecipe, Recipe
 from pythonforandroid.recommendations import (
@@ -89,6 +94,8 @@ class Context:
     bootstrap_build_dir = None
 
     recipe_build_order = None  # Will hold the list of all built recipes
+
+    python_modules = None  # Will hold resolved pure python packages
 
     symlink_bootstrap_files = False  # If True, will symlink instead of copying during build
 
@@ -444,6 +451,12 @@ class Context:
                 # Failed to look up any meaningful name.
                 return False
 
+        # normalize name to remove version tags
+        try:
+            name = Requirement(name).name
+        except Exception:
+            pass
+
         # Try to look up recipe by name:
         try:
             recipe = Recipe.get_recipe(name, self)
@@ -649,6 +662,107 @@ def run_setuppy_install(ctx, project_dir, env=None, arch=None):
             os.remove("._tmp_p4a_recipe_constraints.txt")
 
 
+def is_wheel_platform_independent(whl_name):
+    name, version, build, tags = parse_wheel_filename(whl_name)
+    return all(tag.platform == "any" for tag in tags)
+
+
+def process_python_modules(ctx, modules):
+    """Use pip --dry-run to resolve dependencies and filter for pure-Python packages
+    """
+    modules = list(modules)
+    build_order = list(ctx.recipe_build_order)
+
+    _requirement_names = []
+    processed_modules = []
+
+    for module in modules+build_order:
+        try:
+            # we need to normalize names
+            # eg Requests>=2.0 becomes requests
+            _requirement_names.append(Requirement(module).name)
+        except Exception:
+            # name parsing failed; skip processing this module via pip
+            processed_modules.append(module)
+            if module in modules:
+                modules.remove(module)
+
+    if len(processed_modules) > 0:
+        warning(f'Ignored by module resolver : {processed_modules}')
+
+    # preserve the original module list
+    processed_modules.extend(modules)
+
+    if len(modules) == 0:
+        return modules
+
+    # temp file for pip report
+    fd, path = tempfile.mkstemp()
+    os.close(fd)
+
+    # setup hostpython recipe
+    host_recipe = Recipe.get_recipe("hostpython3", ctx)
+
+    env = environ.copy()
+    _python_path = host_recipe.get_path_to_python()
+    libdir = glob.glob(join(_python_path, "build", "lib*"))
+    env['PYTHONPATH'] = host_recipe.site_dir + ":" + join(
+        _python_path, "Modules") + ":" + (libdir[0] if libdir else "")
+
+    shprint(
+        host_recipe.pip, 'install', *modules,
+        '--dry-run', '--break-system-packages', '--ignore-installed',
+        '--report', path, '-q', _env=env
+    )
+
+    with open(path, "r") as f:
+        report = json.load(f)
+
+    os.remove(path)
+
+    info('Extra resolved pure python dependencies :')
+
+    ignored_str = " (ignored)"
+    # did we find any non pure python package?
+    any_not_pure_python = False
+
+    # just for style
+    info(" ")
+    for module in report["install"]:
+
+        mname = module["metadata"]["name"]
+        mver = module["metadata"]["version"]
+        filename = basename(module["download_info"]["url"])
+        pure_python = True
+
+        if (filename.endswith(".whl") and not is_wheel_platform_independent(filename)):
+            any_not_pure_python = True
+            pure_python = False
+
+        # does this module matches any recipe name?
+        if mname.lower() in _requirement_names:
+            continue
+
+        color = Out_Fore.GREEN if pure_python else Out_Fore.RED
+        ignored = "" if pure_python else ignored_str
+
+        info(
+            f"  {color}{mname}{Out_Fore.WHITE} : "
+            f"{Out_Style.BRIGHT}{mver}{Out_Style.RESET_ALL}"
+            f"{ignored}"
+        )
+
+        if pure_python:
+            processed_modules.append(f"{mname}=={mver}")
+    info(" ")
+
+    if any_not_pure_python:
+        warning("Some packages were ignored because they are not pure Python.")
+        warning("To install the ignored packages, explicitly list them in your requirements file.")
+
+    return processed_modules
+
+
 def run_pymodules_install(ctx, arch, modules, project_dir=None,
                           ignore_setup_py=False):
     """ This function will take care of all non-recipe things, by:
@@ -662,6 +776,10 @@ def run_pymodules_install(ctx, arch, modules, project_dir=None,
     """
 
     info('*** PYTHON PACKAGE / PROJECT INSTALL STAGE FOR ARCH: {} ***'.format(arch))
+
+    # don't run process_python_modules in tests
+    if ctx.recipe_build_order.__class__.__name__ != "Mock":
+        modules = process_python_modules(ctx, modules)
 
     modules = [m for m in modules if ctx.not_has_package(m, arch)]
 

--- a/pythonforandroid/recipe.py
+++ b/pythonforandroid/recipe.py
@@ -979,7 +979,7 @@ class PythonRecipe(Recipe):
         env['LANG'] = "en_GB.UTF-8"
 
         # Binaries made by packages installed by pip
-        self.patch_shebangs(self._host_recipe.local_bin, self.real_hostpython_location)
+        self.patch_shebangs(self._host_recipe.local_bin, self._host_recipe.python_exe)
         env["PATH"] = self._host_recipe.local_bin + ":" + self._host_recipe.site_bin + ":" + env["PATH"]
 
         host_env = self.get_hostrecipe_env(arch)
@@ -1022,10 +1022,9 @@ class PythonRecipe(Recipe):
 
         info('Installing {} into site-packages'.format(self.name))
 
-        hostpython = sh.Command(self.hostpython_location)
         hpenv = env.copy()
         with current_directory(self.get_build_dir(arch.arch)):
-            shprint(hostpython, '-m', 'pip', 'install', '.',
+            shprint(self._host_recipe.pip, 'install', '.',
                     '--compile', '--target',
                     self.ctx.get_python_install_dir(arch.arch),
                     _env=hpenv, *self.setup_extra_args
@@ -1045,8 +1044,7 @@ class PythonRecipe(Recipe):
 
     def install_hostpython_package(self, arch):
         env = self.get_hostrecipe_env(arch)
-        real_hostpython = sh.Command(self.real_hostpython_location)
-        shprint(real_hostpython, '-m', 'pip', 'install', '.',
+        shprint(self._host_recipe.pip, 'install', '.',
                 '--compile',
                 '--root={}'.format(self._host_recipe.site_root),
                 _env=env, *self.setup_extra_args)
@@ -1075,8 +1073,7 @@ class PythonRecipe(Recipe):
             pip_options.append("--upgrade")
         # Use system's pip
         pip_env = self.get_hostrecipe_env()
-        pip_env["HOME"] = "/tmp"
-        shprint(sh.Command(self.real_hostpython_location), "-m", "pip", *pip_options, _env=pip_env)
+        shprint(self._host_recipe.pip, *pip_options, _env=pip_env)
 
     def restore_hostpython_prerequisites(self, packages):
         _packages = []
@@ -1270,10 +1267,14 @@ class PyProjectRecipe(PythonRecipe):
         return env
 
     def get_wheel_platform_tag(self, arch):
+        # https://peps.python.org/pep-0738/#packaging
+        # official python only supports 64 bit:
+        # android_21_arm64_v8a
+        # android_21_x86_64
         return f"android_{self.ctx.ndk_api}_" + {
-            "armeabi-v7a": "arm",
-            "arm64-v8a": "aarch64",
+            "arm64-v8a": "arm64_v8a",
             "x86_64": "x86_64",
+            "armeabi-v7a": "arm",
             "x86": "i686",
         }[arch.arch]
 

--- a/pythonforandroid/recipes/hostpython3/__init__.py
+++ b/pythonforandroid/recipes/hostpython3/__init__.py
@@ -36,7 +36,7 @@ class HostPython3Recipe(Recipe):
         :class:`~pythonforandroid.python.HostPythonRecipe`
     '''
 
-    version = '3.14.0'
+    version = '3.14.2'
 
     url = 'https://github.com/python/cpython/archive/refs/tags/v{version}.tar.gz'
     '''The default url to download our host python recipe. This url will
@@ -45,6 +45,8 @@ class HostPython3Recipe(Recipe):
     build_subdir = 'native-build'
     '''Specify the sub build directory for the hostpython3 recipe. Defaults
     to ``native-build``.'''
+
+    patches = ["fix_ensurepip.patch"]
 
     @property
     def _exe_name(self):
@@ -113,6 +115,33 @@ class HostPython3Recipe(Recipe):
             f"usr/local/lib/python{p_version.major}.{p_version.minor}/site-packages/"
         )
 
+    @property
+    def _pip(self):
+        return join(self.local_bin, "pip3")
+
+    @property
+    def pip(self):
+        return sh.Command(self._pip)
+
+    def fix_pip_shebangs(self):
+
+        if not os.path.exists(self.local_bin):
+            return
+
+        for filename in os.listdir(self.local_bin):
+            if not filename.startswith("pip"):
+                continue
+
+            pip_path = os.path.join(self.local_bin, filename)
+
+            with open(pip_path, "rb") as file:
+                file_lines = file.read().splitlines()
+
+            file_lines[0] = f"#!{self.python_exe}".encode()
+
+            with open(pip_path, "wb") as file:
+                file.write(b"\n".join(file_lines) + b"\n")
+
     def build_arch(self, arch):
         env = self.get_recipe_env(arch)
 
@@ -160,11 +189,14 @@ class HostPython3Recipe(Recipe):
 
         ensure_dir(self.site_root)
         self.ctx.hostpython = self.python_exe
+
         if build_configured:
+
             shprint(
                 sh.Command(self.python_exe), "-m", "ensurepip", "--root", self.site_root, "-U",
-                _env={"HOME": "/tmp"}
+                _env={"HOME": "/tmp", "PATH": self.local_bin}
             )
+            self.fix_pip_shebangs()
 
 
 recipe = HostPython3Recipe()

--- a/pythonforandroid/recipes/hostpython3/fix_ensurepip.patch
+++ b/pythonforandroid/recipes/hostpython3/fix_ensurepip.patch
@@ -1,0 +1,20 @@
+diff '--color=auto' -uNr cpython-3.14.0/Lib/ensurepip/__init__.py cpython-3.14.0.mod/Lib/ensurepip/__init__.py
+--- cpython-3.14.0/Lib/ensurepip/__init__.py	2025-10-07 15:04:52.000000000 +0530
++++ cpython-3.14.0.mod/Lib/ensurepip/__init__.py	2025-12-20 18:18:13.884914683 +0530
+@@ -69,7 +69,15 @@
+     code = f"""
+ import runpy
+ import sys
+-sys.path = {additional_paths or []} + sys.path
++
++# tell ensurepip to ignore site-packages
++
++paths = []
++for path in sys.path:
++    if "site-packages" not in path:
++        paths.append(path)
++
++sys.path = {additional_paths or []} + paths
+ sys.argv[1:] = {args}
+ runpy.run_module("pip", run_name="__main__", alter_sys=True)
+ """

--- a/pythonforandroid/recipes/python3/__init__.py
+++ b/pythonforandroid/recipes/python3/__init__.py
@@ -54,7 +54,7 @@ class Python3Recipe(TargetPythonRecipe):
         :class:`~pythonforandroid.python.GuestPythonRecipe`
     '''
 
-    version = '3.14.0'
+    version = '3.14.2'
     _p_version = Version(version)
     url = 'https://github.com/python/cpython/archive/refs/tags/v{version}.tar.gz'
     name = 'python3'
@@ -78,6 +78,7 @@ class Python3Recipe(TargetPythonRecipe):
 
     if _p_version.minor >= 14:
         patches.append('patches/3.14_armv7l_fix.patch')
+        patches.append('patches/3.14_fix_remote_debug.patch')
 
     if shutil.which('lld') is not None:
         if _p_version.minor == 7:

--- a/pythonforandroid/recipes/python3/patches/3.14_fix_remote_debug.patch
+++ b/pythonforandroid/recipes/python3/patches/3.14_fix_remote_debug.patch
@@ -1,0 +1,28 @@
+diff '--color=auto' -uNr cpython-3.14.2/Modules/_remote_debugging_module.c cpython-3.14.2.mod/Modules/_remote_debugging_module.c
+--- cpython-3.14.2/Modules/_remote_debugging_module.c	2025-12-05 22:19:16.000000000 +0530
++++ cpython-3.14.2.mod/Modules/_remote_debugging_module.c	2025-12-13 20:22:44.011497868 +0530
+@@ -812,7 +812,9 @@
+         PyErr_SetString(PyExc_RuntimeError, "Failed to find the AsyncioDebug section in the process.");
+         _PyErr_ChainExceptions1(exc);
+     }
+-#elif defined(__linux__)
++
++// https://github.com/python/cpython/commit/1963e701001839389cfb1b11d803b0743f4705d7
++#elif defined(__linux__) && HAVE_PROCESS_VM_READV
+     // On Linux, search for asyncio debug in executable or DLL
+     address = search_linux_map_for_section(handle, "AsyncioDebug", "_asyncio.cpython");
+     if (address == 0) {
+diff '--color=auto' -uNr cpython-3.14.2/Python/remote_debug.h cpython-3.14.2.mod/Python/remote_debug.h
+--- cpython-3.14.2/Python/remote_debug.h	2025-12-05 22:19:16.000000000 +0530
++++ cpython-3.14.2.mod/Python/remote_debug.h	2025-12-13 20:23:27.917518543 +0530
+@@ -881,7 +881,9 @@
+             handle->pid);
+         _PyErr_ChainExceptions1(exc);
+     }
+-#elif defined(__linux__)
++
++// https://github.com/python/cpython/commit/1963e701001839389cfb1b11d803b0743f4705d7
++#elif defined(__linux__) && HAVE_PROCESS_VM_READV
+     // On Linux, search for 'python' in executable or DLL
+     address = search_linux_map_for_section(handle, "PyRuntime", "python");
+     if (address == 0) {

--- a/tests/test_build.py
+++ b/tests/test_build.py
@@ -5,7 +5,7 @@ from unittest import mock
 import jinja2
 
 from pythonforandroid.build import (
-    Context, RECOMMENDED_TARGET_API, run_pymodules_install,
+    Context, RECOMMENDED_TARGET_API, run_pymodules_install, process_python_modules
 )
 from pythonforandroid.archs import ArchARMv7_a, ArchAarch_64
 
@@ -17,7 +17,7 @@ class TestBuildBasic(unittest.TestCase):
         Makes sure the `run_pymodules_install()` doesn't crash when the
         `project_dir` optional parameter is None, refs #1898
         """
-        ctx = mock.Mock()
+        ctx = mock.Mock(recipe_build_order=[])
         ctx.archs = [ArchARMv7_a(ctx), ArchAarch_64(ctx)]
         modules = []
         project_dir = None
@@ -26,8 +26,19 @@ class TestBuildBasic(unittest.TestCase):
         assert m_info.call_args_list[-1] == mock.call(
             'No Python modules and no setup.py to process, skipping')
 
+    def test_python_module_parser(self):
+        ctx = mock.Mock(recipe_build_order=[])
+        ctx.archs = [ArchARMv7_a(ctx), ArchAarch_64(ctx)]
+        # should not alter original module name (like with adding version number)
+        assert "kivy_garden.frostedglass" in process_python_modules(ctx, ["kivy_garden.frostedglass"])
+
+        # should skip urls and other unsupported format
+        modules = ["https://example.com/some.zip", "git+https://github.com/kivy/python-for-android@develop"]
+        result = process_python_modules(ctx, modules)
+        assert modules == result
+
     def test_strip_if_with_debug_symbols(self):
-        ctx = mock.Mock()
+        ctx = mock.Mock(recipe_build_order=[])
         ctx.python_recipe.major_minor_version_string = "3.6"
         ctx.get_site_packages_dir.return_value = "test-doesntexist"
         ctx.build_dir = "nonexistant_directory"

